File: 0386-Lexicographical Numbers/0386-Lexicographical Numbers.cpp
class Solution {
public:
    vector<int> lexicalOrder(int n) {
        vector<int> result(n);
        int curr = 1;
        for (int i = 0; i < n; ++i) {
            result[i] = curr;
            if (10 * curr <= n) {
                curr *= 10;
            }
            else {
                if (curr >= n) {
                    curr /= 10;
                }
                ++curr;
                while (curr % 10 == 0) {
                    curr /= 10;
                }
            }
        }
        
        return result;
    }
};

File: 0301-0400/0386-Lexicographical Numbers/0386-Lexicographical Numbers.py
class Solution:
    def lexicalOrder(self, n: int) -> List[int]:
        result = [0] * n
        curr = 1
        for i in range(n):
            result[i] = curr
            if curr * 10 <= n:
                curr *= 10
            else:
                if curr >= n:
                    curr //= 10
                curr += 1
                while curr % 10 == 0:
                    curr //= 10
        return result
